fix issues related to prometheus relabel configs when target allocator is enabled

anunarapureddy opened the pull request:
Related to following issues: #[1623](https://github.com/open-telemetry/opentelemetry-operator/issues/1623), #[1622](https://github.com/open-telemetry/opentelemetry-operator/issues/1622), #[958](https://github.com/open-telemetry/opentelemetry-operator/issues/958)

### Issue: Collector and Target Allocator pods fail to get created in certain scenarios when Target Allocator is enabled and Prometheus configuration contains labelmap, labeldrop, or labelkeep actions. 

### Cause:

The issue manifests itself in the following scenarios.

1. First, when the configMap is built by the operator, the unmarshalling logic calls the Prometheus relabel config validation logic ([relabel.go](https://github.com/prometheus/prometheus/blob/3923e83413075a17c5a10739696a879765a05542/model/relabel/relabel.go#L103)), which includes the additional defaults, including the $1 replacement field, appended to the original Prometheus configuration. Upon initialization, the collector pod loads the configMap and interprets $1 as an environment variable, leading to its replacement with an empty string. As a result, the same Prometheus relabel config validation logic fails when the config is unmarshalled in the collector code, as the replacement key is now replaced with an empty string due to env var substitution. 

2. The second scenario occurs when the Prometheus configuration explicitly includes $$ for labelmap, labeldrop, labelkeep or other actions that undergo regex validation in the Prometheus validation code ([relabel.go](https://github.com/prometheus/prometheus/blob/3923e83413075a17c5a10739696a879765a05542/model/relabel/relabel.go#L126)). In this case, when the operator creates the collector configMap, the regex validation fails, causing the configMap to get created with an empty yaml file.

3. In the third scenario, when Prometheus configuration explicitly includes $$ for labelmap, labeldrop, labelkeep or other actions, the TA ConfigMap is created with the $$ signs. Since the TA does not support environment variable substitution, the unmarshalling logic during the ConfigMap mount process in the TA pod will perform regex validation and cause the TA pod creation to fail.

### Resolution:

Below is the guideline from OpenTelemetry documentation:

> Note: Since the collector configuration supports env variable substitution $ characters in your prometheus configuration are interpreted as environment variables. If you want to use $ characters in your prometheus configuration, you must escape them using $$.

Key points considered in order to address the issue:

- To use "$" characters in Prometheus configuration, they must be escaped using "$$".
- The collector configuration supports environment variable substitution, but the TA does not.
- The ConfigMap for the collector pod should have "$$" to prevent env var substitution, unless the user intentionally wants it to happen.
- The TA ConfigMap should have a single "$", as it does not support env var substitution.

The operator was updated to modify the YAML file directly instead of going through the Prometheus unmarshalling logic to update the SD config. The TA ConfigMap creation process was modified to replace "$$" signs with a "$" to address the third scenario. These changes ensure that both the Collector and TA pods are created correctly, even when the Prometheus configuration includes labelmap, labeldrop, labelkeep, or other actions that undergo regex validation.

cc: @alolita @jaronoff97

anunarapureddy commented:
@open-telemetry/operator-maintainers As a first-time contributor, I need approval from a maintainer to run workflows. Would someone be able to assist me with obtaining approval?

kristinapathak commented:
I've looked and thought about this but mostly have follow up questions.

For the issue with including `$1` - my understanding is this is an issue due to the TA being enabled and adding back `$1`'s which are [not supported](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/prometheusreceiver#getting-started) in the collector configuration:

>Note: Since the collector configuration supports env variable substitution $ characters in your prometheus configuration are interpreted as environment variables. If you want to use $ characters in your prometheus configuration, you must escape them using $$.

While I wish they could be supported, I don't think we should be using the operator to make `$` variables work (which I believe this PR is doing).  My recommendation is instead to modify [`DefaultRelabelConfig`](https://github.com/prometheus/prometheus/blob/3923e83413075a17c5a10739696a879765a05542/model/relabel/relabel.go#L31) to no longer set `Replacement`.  This is not ideal, but because it's a public var, we can set it in the operator code to our desired default config and the prometheus code will use the new value.  Then `$1`'s will no longer be added back into the scrape config when the TA is enabled.

The second issue I see being addressed in this PR is prometheus not validating regex with `$$`.  Is this only with TA enabled or regardless of TA enablement?  And is this due to the code change of `$`->`$$`? I can't find anything in the prometheus code that explains why this wouldn't work.

I looked through the code in the prometheus receiver to find somewhere that `$$` was checked or treated specially but couldn't find anything, so my assumption is that `$$` is considered valid in the scrape config, just doesn't pass the regex checks for some reason.  In other words, the same scrape config with `$$` passed in the operator which causes a problem could be passed in a collector config directly with no problem.  Is that true?  I plan to look into this more but wanted to be sure to provide timely feedback.

My general takeaway is that there are multiple issues here, and I would rather detangle and solve them separately, if possible.  Is there any other issue being addressed here that I've missed?

anunarapureddy commented:
@open-telemetry/operator-maintainers As a first-time contributor, I need approval from a maintainer to run workflows. Would someone be able to assist me with obtaining approval?

anunarapureddy commented:
@kristinapathak, I have made changes to the pull request and also updated the description. Can you please review the latest code and provide feedback so that we can iterate on the changes? I was brainstorming with @jaronoff97 and came up with an alternative approach, which is reflected in the updated code.

anunarapureddy commented:
@kristinapathak, I have made changes to the pull request and also updated the description. Can you please review the latest code and provide feedback so that we can iterate on the changes?

jaronoff97 commented on pkg/collector/testdata/relabel_config_expected_with_sd_config.yaml:
```diff
@@ -0,0 +1,57 @@
+          - label1
+        - action: replace
+          regex: (.*)
+          replacement: $$1_$2
```
should the `$2` be escaped as `$$2`? This would be what we tell a collector user to do, no?

jaronoff97 replied:
i.e. does the first `$` escape just `$1` or does it escape the entire `$1_$2`?

anunarapureddy replied:
@jaronoff97, the code path for constructing the collector configMap does not escape any dollar characters ('$') in the user-specified configuration. Instead, it updates the HTTP SD service configuration by directly modifying the promConfigMap, without the need for unmarshalling and updating the configuration. It is expected that users will properly escape '$' characters in their Prometheus configuration. This approach allows us to support environment variable substitution as well. For instance, in the example above, if a user entered "$$1_$2" in the prometheus configuration, the config map would be created with that value. The "$$1" part will be processed by the regular expression match, while the "$2" part will be replaced by the corresponding environment variable. The fix ensures that there are no problems when the users specify "$$" in the prometheus configuration (as per the OTel guideline).

jaronoff97 replied:
i see, so in this case `$$1_$2` for a regex match of 'thing' would be `thing_` if $2 is unset? I guess my question is, should we also escape the $2 in our test?

jaronoff97 commented on pkg/collector/reconcile/configmap_test.go:
```diff
@@ -94,23 +94,11 @@ receivers:
+        - url: http://test-targetallocator:80/jobs/otel-collector/targets?collector_id=$POD_NAME
+        job_name: otel-collector
         scrape_interval: 10s
-        scrape_timeout: 10s
-        metrics_path: /metrics
-        scheme: http
-        follow_redirects: true
-        enable_http2: true
-        http_sd_configs:
-        - follow_redirects: false
-          enable_http2: false
-          url: http://test-targetallocator:80/jobs/otel-collector/targets?collector_id=$POD_NAME
```
Why are we removing these lines?

anunarapureddy replied:
@jaronoff97 These are additional global defaults that are included in the prometheus configuration when it is unmarshalled. Since we are no longer unmarshalling the configuration, these defaults had to be removed. The [original configuration](https://github.com/open-telemetry/opentelemetry-operator/blob/main/pkg/collector/reconcile/test.yaml) does not contain these defaults.

jaronoff97 commented on pkg/targetallocator/adapters/config_to_prom_config.go:
```diff
@@ -57,3 +68,170 @@ func ConfigToPromConfig(cfg string) (map[interface{}]interface{}, error) {
+				return nil, errorNotAString("replacement")
+			}
+
+			relabelConfig["replacement"] = strings.ReplaceAll(replacement, "$$", "$")
```
are we unable to just do the replace all on the string configuration we're given? Do we need to only run it on relabel_configs and metric_relabel_configs?

jaronoff97 replied:
@kristinapathak thinks we should maybe try and just regex this change as well.

jaronoff97 commented on tests/e2e/targetallocator-features/00-install.yaml:
```diff
@@ -53,10 +53,16 @@ spec:
+            static_configs:
+            - targets: [ '0.0.0.0:8888' ]
+            relabel_configs:
+            - regex: __meta_kubernetes_node_label_(.+)
```
for completeness, can we also add in a labelkeep action here too?

anunarapureddy replied:
@jaronoff97 Done!

jaronoff97 commented on pkg/collector/reconcile/config_replace.go:
```diff
@@ -44,55 +41,67 @@ type Config struct {
-				URL: fmt.Sprintf("http://%s:80/jobs/%s/targets?collector_id=$POD_NAME", naming.TAService(instance), escapedJob),
-			},
+		// yaml marshaling/unsmarshaling is preferred because of the problems associated with the conversion of map to a struct using mapstructure
+		promCfg, marshalErr := yaml.Marshal(promCfgMap)
```
to confirm my understanding of this – does doing the marshal and then the unmarshal prevent the defaultrelabelconfig from being added and thus removing [this error](https://github.com/prometheus/prometheus/blob/main/model/relabel/relabel.go#LL148C49-L148C49) from being thrown?

jaronoff97 replied:
or is that going to be in a separate PR?

anunarapureddy replied:
@jaronoff97 I had initially planned to handle this separately, but later decided to include it in the same PR to ensure a clear and coherent thought process. I pushed the latest changes to handle this scenario as well.

jaronoff97 reviewed (commented):
Overall, @kristinapathak and I definitely like this version more than the massaging you had previously. It's easier to understand and the overall design is more extendable. We had a few thoughts/questions to confirm what we had discussed. Thank you again for your contribution 🙇

jaronoff97 reviewed (commented):
Overall, @kristinapathak and I definitely like this version more than the massaging you had previously. It's easier to understand and the overall design is more extendable. We had a few thoughts/questions to confirm what we had discussed. Thank you again for your contribution 🙇

TylerHelmuth commented:
@anunarapureddy grab the latest changes from main to deal with the failing autoscale e2e test.

kristinapathak commented on pkg/targetallocator/adapters/config_to_prom_config.go:
```diff
@@ -57,3 +68,208 @@ func ConfigToPromConfig(cfg string) (map[interface{}]interface{}, error) {
+
+	// we don't need the scrape configs here anymore with target allocator enabled
+	// Remove the scrape_configs key from the map
+	delete(prometheusCfg, "scrape_configs")
```
Can you also add an explanation for this in the function comment?

anunarapureddy replied:
Done.

kristinapathak commented on pkg/targetallocator/adapters/config_to_prom_config.go:
```diff
@@ -57,3 +68,208 @@ func ConfigToPromConfig(cfg string) (map[interface{}]interface{}, error) {
+
+// UnescapeDollarSignsInPromConfig replaces "$$" with "$" in the "replacement" fields of
+// both "relabel_configs" and "metric_relabel_configs" in a Prometheus configuration file.
+func UnescapeDollarSignsInPromConfig(cfg string) (map[interface{}]interface{}, error) {
```
What is the test coverage like on these new functions? It seems like there are a lot of `if` statements present but very few test cases.

anunarapureddy replied:
Adding test cases for missing and invalid configurations as well.

kristinapathak reviewed (commented):
This looks great! Left a few very minor suggestions/questions.

jaronoff97 commented:
@anunarapureddy we just added some more e2e tests to help us catch more of these in the future, and there's already one for labeldropping that you just need to uncomment to enable. ([PR](https://github.com/open-telemetry/opentelemetry-operator/pull/1762)) Could you rebase and uncomment that test when you get a chance as well?

anunarapureddy commented on pkg/targetallocator/adapters/config_to_prom_config.go:
```diff
@@ -57,3 +68,208 @@ func ConfigToPromConfig(cfg string) (map[interface{}]interface{}, error) {
+	for _, config := range scrapeConfigs {
+		scrapeConfig, ok := config.(map[interface{}]interface{})
+		if !ok {
+			return nil, errorNotAMap("scrape_config")
```
Done.

anunarapureddy commented on pkg/targetallocator/adapters/config_to_prom_config.go:
```diff
@@ -57,3 +68,208 @@ func ConfigToPromConfig(cfg string) (map[interface{}]interface{}, error) {
+				continue
+			}
+			if sdRegex.MatchString(keyStr) {
+				delete(scrapeConfig, key)
```
Done.

anunarapureddy commented:
@jaronoff97 Uncommented the "labeldrop" e2e test case and added additional unit test cases to cover missing or invalid configurations.

jaronoff97 commented:
Awesome, thank you! running those now and will give another review today as well once they pass 🤞

jaronoff97 commented:
looks like a single linter bug @anunarapureddy

anunarapureddy commented:
@jaronoff97 Seems like the linter gave up because it took too long. I checked the code on my own machine, and I didn't spot any lint failures. Everything looks good from my end!

jaronoff97 commented on pkg/collector/reconcile/config_replace.go:
```diff
@@ -36,9 +32,6 @@ type targetAllocator struct {
 	CollectorID string        `yaml:"collector_id"`
-	// HTTPSDConfig is a preference that can be set for the collector's target allocator, but the operator doesn't
-	// care about what the value is set to. We just need this for validation when unmarshalling the configmap.
-	HTTPSDConfig interface{} `yaml:"http_sd_config,omitempty"`
```
i think we still need this? otherwise we'll run in to the unmarshal error about an unknown field.

anunarapureddy replied:
@jaronoff97 Apologies! It seems that this happened during the rebase and conflict resolution process. I have thoroughly reviewed everything now, and it appears to be in good shape.

jaronoff97 reviewed (approved):
just one quick question, otherwise this is looking good to me!

swiatekm-sumo commented on pkg/targetallocator/adapters/config_to_prom_config.go:
```diff
@@ -59,6 +78,201 @@ func ConfigToPromConfig(cfg string) (map[interface{}]interface{}, error) {
+		return nil, err
+	}
+
+	prometheusConfigProperty, ok := prometheus["config"]
+	if !ok {
+		return nil, errorNoComponent("prometheusConfig")
+	}
+
+	prometheusConfig, ok := prometheusConfigProperty.(map[interface{}]interface{})
+	if !ok {
+		return nil, errorNotAMap("prometheusConfig")
+	}
+
+	scrapeConfigsProperty, ok := prometheusConfig["scrape_configs"]
+	if !ok {
+		return nil, errorNoComponent("scrape_configs")
+	}
+
+	scrapeConfigs, ok := scrapeConfigsProperty.([]interface{})
+	if !ok {
+		return nil, errorNotAList("scrape_configs")
+	}
```
I don't think we should error if these don't exist, but rather just exit. Target allocator doesn't require Prometheus configuration.

Maybe this should be done in a separate PR though, as I believe this mirrors the current behaviour. WDYT @jaronoff97 ?

anunarapureddy replied:
@jaronoff97 @swiatekm-sumo I can create a follow up PR to relax this restriction.

swiatekm-sumo commented on pkg/collector/reconcile/configmap.go:
```diff
@@ -103,7 +103,7 @@ func desiredTAConfigMap(params Params) (corev1.ConfigMap, error) {
 	}
 
-	prometheusReceiverConfig, err := ta.ConfigToPromConfig(params.Instance.Spec.Config)
+	prometheusReceiverConfig, err := ta.UnescapeDollarSignsInPromConfig(params.Instance.Spec.Config)
```
I think it's worth adding a comment here as well. Without context it's not clear at all why we'd need to unescape these here.

anunarapureddy replied:
Done. Added a comment with the reason to unescape.

swiatekm-sumo commented on pkg/targetallocator/adapters/config_to_prom_config.go:
```diff
@@ -59,6 +78,201 @@ func ConfigToPromConfig(cfg string) (map[interface{}]interface{}, error) {
+// from the `scrape_configs` section and adds a single `http_sd_configs` configuration.
+// The `http_sd_configs` points to the TA (Target Allocator) endpoint that provides the list of targets for the given job.
+func AddHTTPSDConfigToPromConfig(prometheus map[interface{}]interface{}, taServiceName string) (map[interface{}]interface{}, error) {
+	prometheusConfigProperty, ok := prometheus["config"]
+	if !ok {
+		return nil, errorNoComponent("prometheusConfig")
+	}
+
+	prometheusConfig, ok := prometheusConfigProperty.(map[interface{}]interface{})
+	if !ok {
+		return nil, errorNotAMap("prometheusConfig")
+	}
+
+	scrapeConfigsProperty, ok := prometheusConfig["scrape_configs"]
+	if !ok {
+		return nil, errorNoComponent("scrape_configs")
+	}
+
+	scrapeConfigs, ok := scrapeConfigsProperty.([]interface{})
+	if !ok {
+		return nil, errorNotAList("scrape_configs")
+	}
```
Should we error if the Prometheus config doesn't exist, or just return quietly? See my comment for the TA config unescaping as well.

swiatekm-sumo reviewed (commented):
Looks good to me overall. I think we're being too strict about requiring the Prometheus config to exist, but that can be relaxed in a future PR if need be. Overall, despite the verbosity (should we consider parsing permissively into structs instead?), I think this will greatly improve the reliability of the config transformations.

anunarapureddy commented:
@jaronoff97 @swiatekm-sumo Addressed the review comments and I can work on creating a follow up PR to relax the restrictions on prometheus configuration.

swiatekm-sumo reviewed (approved):
LGTM

anunarapureddy commented:
@bryan-aguilar That's a good suggestion. I can create a follow up PR to update the TA documentation.

bryan-aguilar commented:
> @bryan-aguilar That's a good suggestion. I can create a follow up PR to update the TA documentation.

Do we predict further changes to this functionality after this PR? I worry about documentation drift and a possible loss of context if we don't lump documentation and functional changes in the same PR.

anunarapureddy commented:
> > @bryan-aguilar That's a good suggestion. I can create a follow up PR to update the TA documentation.
> 
> Do we predict further changes to this functionality after this PR? I worry about documentation drift and a possible loss of context if we don't lump documentation and functional changes in the same PR.

Could you please review the Updated the README documentation which reflects the translations being done?

Aneurysm9 commented on README.md:
```diff
@@ -429,28 +436,27 @@ spec:
-            evaluation_interval: 1m
           scrape_configs:
           - job_name: otel-collector
-            honor_timestamps: true
             scrape_interval: 10s
-            scrape_timeout: 10s
-            metrics_path: /metrics
-            scheme: http
-            follow_redirects: true
```
Shouldn't these values be preserved in the configuration given to the collector?  These are scrape-time options that are needed by the receiver.

Aneurysm9 commented on README.md:
````diff
@@ -463,7 +469,26 @@ Behind the scenes, the OpenTelemetry Operator will convert the Collector’s con
 ```
 
-Note how the Operator added a `global` section and a new `http_sd_configs` to the `otel-collector` scrape config, pointing to a Target Allocator instance it provisioned.
+Note how the Operator removes any existing service discovery configurations (e.g., `static_configs`, `file_sd_configs`, etc.) from the `scrape_configs` section and adds a single `http_sd_configs` configuration, pointing to a Target Allocator instance it provisioned.
````
```suggestion
Note how the Operator removes any existing service discovery configurations (e.g., `static_configs`, `file_sd_configs`, etc.) from the `scrape_configs` section and adds an `http_sd_configs` configuration pointing to a Target Allocator instance it provisioned.
```

More than one `http_sd_configs` entry will be present if there is more than one scrape job in the incoming config.  There should be one per job.  The example only had one job, but a real config is likely to have more.

anunarapureddy replied:
@Aneurysm9 The replacement process occurs for each job in the incoming configuration. The code iterates through all the jobs and updates the configuration accordingly. If it would be helpful for clarity, I can include an additional job in the example.

Aneurysm9 commented on pkg/collector/reconcile/config_replace.go:
```diff
@@ -65,42 +63,39 @@ func ReplaceConfig(instance v1alpha1.OpenTelemetryCollector) (string, error) {
-		return "", fmt.Errorf("error unmarshaling YAML: %w", marshalErr)
-	}
+		// type coercion checks are handled in the AddTAConfigToPromConfig method above
+		config["receivers"].(map[interface{}]interface{})["prometheus"] = updPromCfgMap
```
I don't believe it is safe to assume that the receiver ID is always `prometheus`.  There may be multiple prometheus receiver configurations and any or all of them may have a name in addition to the type in their identifier.  This needs to iterate over all entries in the `receivers` map and process any that have an identifier matching `/^prometheus(\/.+)?$/`.

anunarapureddy replied:
@Aneurysm9 The assignment of this logic remains unchanged as it is an existing implementation (https://github.com/open-telemetry/opentelemetry-operator/blob/d6d9f2f628a80858460cddbbee85b4ca9f13fecf/pkg/collector/reconcile/config_replace.go#L99). However, I can certainly address it as an improvement in a follow-up PR.

kristinapathak reviewed (approved):
This is awesome! Thanks for putting in so much time and effort into getting `$`s working 🙂

bryan-aguilar reviewed (approved):
Thank you for adding the documentation!

jaronoff97 commented:
once tests pass, will merge this in :)
